File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        result = []
        frontier = [root]
        
        while frontier:
            newFrontier = []
            currLvl = []
            
            for node in frontier:
                currLvl.append(node.val)
                
                if node.left:
                    newFrontier.append(node.left)
                if node.right:
                    newFrontier.append(node.right)
            frontier = newFrontier
            result.append(currLvl)
            
        return result

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root == nullptr) {
            return {};
        }
        vector<vector<int>> result;
        vector<TreeNode*> frontier {root};
        
        while (!frontier.empty()) {
            vector<TreeNode*> newFrontier;
            vector<int> curr;
            
            for (auto node: frontier) {
                curr.push_back(node->val);
                if (node->left != nullptr) {
                    newFrontier.push_back(node->left);
                }
                if (node->right != nullptr) {
                    newFrontier.push_back(node->right);
                }
            }
            swap(frontier, newFrontier);
            result.push_back(curr);
        }
        return result;
    }
};